Issue with Strimzi discovery when no Strimzi is installed

## This is a...

<pre><code>
[ ] Feature request
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  <!-- Please search GitHub for a similar issue or PR before submitting -->
[ ] Documentation issue or request
</code></pre>

## Description

I saw this on the staging cluster, I think the root cause might be that the Strimzi is not installed there. The call in the `meta` pod fails with:

```
2020-01-08 10:26:22.931 ERROR 1 --- [  XNIO-1 task-1] i.s.c.meta.VerifierExceptionMapper       : Exception while handling request: POST /api/v1/connectors/kafka/properties/meta

io.syndesis.common.util.SyndesisServerException: Error while performing the call to https://172.30.0.1:443/apis/kafka.strimzi.io/v1beta1/kafkas/ . Response code: 403. Unable to fetch and process metadata
	at io.syndesis.connector.support.verifier.api.MetadataRetrieval.handle(MetadataRetrieval.java:51) ~[connector-support-verifier-2.0-SNAPSHOT.jar!/:2.0-SNAPSHOT]
	at io.syndesis.connector.meta.v1.ConnectorEndpoint.properties(ConnectorEndpoint.java:79) ~[classes!/:2.0-SNAPSHOT]
...
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Error while performing the call to https://172.30.0.1:443/apis/kafka.strimzi.io/v1beta1/kafkas/ . Response code: 403
	at io.fabric8.kubernetes.client.dsl.internal.RawCustomResourceOperationsImpl.makeCall(RawCustomResourceOperationsImpl.java:582) ~[kubernetes-client-4.6.1.jar!/:na]
	at io.fabric8.kubernetes.client.dsl.internal.RawCustomResourceOperationsImpl.list(RawCustomResourceOperationsImpl.java:338) ~[kubernetes-client-4.6.1.jar!/:na]
	at io.syndesis.connector.kafka.KafkaMetaDataRetrieval.fetchProperties(KafkaMetaDataRetrieval.java:92) ~[connector-kafka-2.0-SNAPSHOT.jar!/:2.0-SNAPSHOT]
	at io.syndesis.connector.meta.v1.ConnectorEndpoint.properties(ConnectorEndpoint.java:74) ~[classes!/:2.0-SNAPSHOT]
```

That cascaded to the `server` pod as:

```
2020-01-08 15:08:00.608 ERROR 1 --- [  XNIO-1 task-6] .s.e.v.h.e.SyndesisServerExceptionMapper : Internal Server Exception. null

java.lang.NullPointerException: null
	at io.syndesis.server.endpoint.v1.handler.connection.ConnectorHandler.enrichWithDynamicProperties(ConnectorHandler.java:166) ~[server-endpoint-2.0-SNAPSHOT.jar!/:2.0-SNAPSHOT]
	at io.syndesis.server.endpoint.v1.handler.connection.ConnectorHandler.get(ConnectorHandler.java:137) ~[server-endpoint-2.0-SNAPSHOT.jar!/:2.0-SNAPSHOT]
```

--- a/app/server/endpoint/pom.xml
+++ b/app/server/endpoint/pom.xml
@@ -204,6 +204,11 @@
     </dependency>
 
     <dependency>
+      <groupId>com.fasterxml.jackson.jaxrs</groupId>
+      <artifactId>jackson-jaxrs-json-provider</artifactId>
+    </dependency>
+
+    <dependency>
       <groupId>org.jboss.spec.javax.ws.rs</groupId>
       <artifactId>jboss-jaxrs-api_2.1_spec</artifactId>
     </dependency>

--- a/app/server/endpoint/src/main/java/io/syndesis/server/endpoint/v1/handler/connection/ConnectorHandler.java
+++ b/app/server/endpoint/src/main/java/io/syndesis/server/endpoint/v1/handler/connection/ConnectorHandler.java
@@ -25,6 +25,8 @@ import javax.ws.rs.Path;
 import javax.ws.rs.PathParam;
 import javax.ws.rs.Produces;
 import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status.Family;
 import javax.ws.rs.core.UriInfo;
 import java.io.BufferedInputStream;
 import java.io.IOException;
@@ -41,7 +43,6 @@ import java.util.Optional;
 import java.util.stream.Collectors;
 
 import com.fasterxml.jackson.databind.JsonNode;
-import com.fasterxml.jackson.databind.ObjectMapper;
 import io.swagger.annotations.Api;
 import io.swagger.annotations.ApiParam;
 import io.syndesis.common.model.Kind;
@@ -70,8 +71,6 @@ import io.syndesis.server.inspector.Inspectors;
 import io.syndesis.server.verifier.MetadataConfigurationProperties;
 import io.syndesis.server.verifier.Verifier;
 import org.jboss.resteasy.annotations.providers.multipart.MultipartForm;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 import org.springframework.context.ApplicationContext;
 import org.springframework.stereotype.Component;
 
@@ -79,9 +78,6 @@ import org.springframework.stereotype.Component;
 @Api(value = "connectors")
 @Component
 public class ConnectorHandler extends BaseHandler implements Lister<Connector>, Getter<Connector>, Updater<Connector>, Deleter<Connector> {
-
-    private static final Logger LOG = LoggerFactory.getLogger(ConnectorHandler.class);
-    private final ObjectMapper mapper = new ObjectMapper();
 
     private final ApplicationContext applicationContext;
     private final IconDao iconDao;
@@ -133,12 +129,7 @@ public class ConnectorHandler extends BaseHandler implements Lister<Connector>, 
 
         // Retrieve dynamic properties, if connector is dynamic
         if (connector.getTags().contains("dynamic")) {
-            try {
-                connector = enrichWithDynamicProperties(connector);
-            } catch (IOException e) {
-                LOG.error("Issue while enriching with metadata", e);
-                // We can go ahead, just ignoring the dynamic metadata
-            }
+            connector = enrichWithDynamicProperties(connector);
         }
 
         final Optional<String> connectorGroupId = connector.getConnectorGroupId();
@@ -158,13 +149,23 @@ public class ConnectorHandler extends BaseHandler implements Lister<Connector>, 
      * @param connector
      * @return an enriched {@link Connector}
      */
-    private Connector enrichWithDynamicProperties(Connector connector) throws IOException {
-        Map<String, ConfigurationProperty> dynamicProperties = new HashMap<>();
-        String dynamicPropertiesFromMeta = (String) this.properties(connector.getId().get())
-            .enrichWithDynamicProperties(connector.getId().get(), null)
-            .getEntity();
-        Iterator<Map.Entry<String, JsonNode>> iterator = mapper.readTree(dynamicPropertiesFromMeta).get("properties").fields();
-        while(iterator.hasNext()){
+    Connector enrichWithDynamicProperties(Connector connector) {
+        final String connectorId = connector.getId().get();
+        final ConnectorPropertiesHandler propertiesHandler = properties(connectorId);
+
+        final JsonNode jsonResponse;
+        try (Response response = propertiesHandler.enrichWithDynamicProperties(connectorId, null)) {
+            if (response.getStatusInfo().getFamily() != Family.SUCCESSFUL) {
+                return connector;
+            }
+
+            jsonResponse = response.readEntity(JsonNode.class);
+        }
+
+        final Iterator<Map.Entry<String, JsonNode>> iterator = jsonResponse.path("properties").fields();
+        final Map<String, ConfigurationProperty> dynamicProperties = new HashMap<>();
+
+        while (iterator.hasNext()) {
             Map.Entry<String, JsonNode> next = iterator.next();
             String propertyName = next.getKey();
             JsonNode property = next.getValue();

--- a/app/server/endpoint/src/main/java/io/syndesis/server/endpoint/v1/handler/connection/ConnectorPropertiesHandler.java
+++ b/app/server/endpoint/src/main/java/io/syndesis/server/endpoint/v1/handler/connection/ConnectorPropertiesHandler.java
@@ -15,39 +15,42 @@
  */
 package io.syndesis.server.endpoint.v1.handler.connection;
 
+import java.util.Map;
+
 import javax.ws.rs.POST;
 import javax.ws.rs.PathParam;
 import javax.ws.rs.Produces;
+import javax.ws.rs.client.Client;
 import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.WebTarget;
 import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.Response;
-import javax.ws.rs.core.Response.Status;
-import java.util.Map;
 
 import io.swagger.annotations.Api;
 import io.syndesis.server.verifier.MetadataConfigurationProperties;
-import org.jboss.resteasy.client.jaxrs.ResteasyClient;
+
 import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
-import org.jboss.resteasy.client.jaxrs.ResteasyWebTarget;
 
 @Api(value = "properties")
 public class ConnectorPropertiesHandler {
-    private final MetadataConfigurationProperties config;
+    private final String metaPropertiesUrl;
 
     public ConnectorPropertiesHandler(final MetadataConfigurationProperties config) {
-        this.config = config;
+        metaPropertiesUrl = String.format("http://%s/api/v1/connectors/%%s/properties/meta", config.getService());
     }
 
     @POST
     @Produces(MediaType.APPLICATION_JSON)
     public Response enrichWithDynamicProperties(@PathParam("id") final String connectorId, final Map<String, Object> props) {
         // TODO replace properly with circuit breaker
-        String metadataUrl = String.format("http://%s/api/v1/connectors/%s/properties/meta", config.getService(), connectorId);
-        ResteasyClient client = new ResteasyClientBuilder().build();
-        ResteasyWebTarget target = client.target(metadataUrl);
-        try (Response response = target.request().post(Entity.entity(props, "application/json"));) {
-            final Status status = Status.OK;
-            return Response.status(status).entity(response.readEntity(String.class)).build();
-        }
+        String metadataUrl = String.format(metaPropertiesUrl, connectorId);
+        Client client = createClient();
+        WebTarget target = client.target(metadataUrl);
+
+        return target.request().post(Entity.entity(props, MediaType.APPLICATION_JSON_TYPE));
+    }
+
+    Client createClient() {
+        return new ResteasyClientBuilder().build();
     }
 }

--- a/app/server/endpoint/src/test/java/io/syndesis/server/endpoint/v1/handler/connection/ConnectorHandlerTest.java
+++ b/app/server/endpoint/src/test/java/io/syndesis/server/endpoint/v1/handler/connection/ConnectorHandlerTest.java
@@ -19,11 +19,15 @@ import static javax.ws.rs.core.HttpHeaders.CONTENT_LENGTH;
 import static javax.ws.rs.core.HttpHeaders.CONTENT_TYPE;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
+import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashSet;
@@ -33,12 +37,20 @@ import java.util.List;
 import javax.imageio.ImageIO;
 import javax.imageio.ImageReader;
 import javax.imageio.stream.ImageInputStream;
+import javax.validation.constraints.NotNull;
+import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.Response;
 import javax.ws.rs.core.StreamingOutput;
 
 import io.syndesis.server.verifier.MetadataConfigurationProperties;
+
+import org.jboss.resteasy.client.jaxrs.internal.ClientConfiguration;
+import org.jboss.resteasy.client.jaxrs.internal.ClientResponse;
+import org.jboss.resteasy.spi.ResteasyProviderFactory;
 import org.junit.Test;
 import org.springframework.context.ApplicationContext;
+
+import com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider;
 
 import io.syndesis.server.credential.Credentials;
 import io.syndesis.server.dao.file.FileDataManager;
@@ -49,8 +61,10 @@ import io.syndesis.server.inspector.Inspectors;
 import io.syndesis.common.model.ListResult;
 import io.syndesis.common.model.action.ConnectorAction;
 import io.syndesis.common.model.action.ConnectorDescriptor;
+import io.syndesis.common.model.connection.ConfigurationProperty;
 import io.syndesis.common.model.connection.Connection;
 import io.syndesis.common.model.connection.Connector;
+import io.syndesis.common.model.connection.ConfigurationProperty.PropertyValue;
 import io.syndesis.common.model.integration.Flow;
 import io.syndesis.common.model.integration.Integration;
 import io.syndesis.common.model.integration.Step;
@@ -85,8 +99,8 @@ public class ConnectorHandlerTest {
 
     private static final MetadataConfigurationProperties NO_METADATA_CONFIGURATION_PROPERTIES = null;
 
-    private final io.syndesis.server.endpoint.v1.handler.connection.ConnectorHandler handler =
-        new io.syndesis.server.endpoint.v1.handler.connection.ConnectorHandler(dataManager, NO_VERIFIER, NO_CREDENTIALS, NO_INSPECTORS, NO_STATE,
+    private final ConnectorHandler handler =
+        new ConnectorHandler(dataManager, NO_VERIFIER, NO_CREDENTIALS, NO_INSPECTORS, NO_STATE,
             NO_ENCRYPTION_COMPONENT, applicationContext, NO_ICON_DAO, NO_EXTENSION_DATA_MANAGER,
             NO_METADATA_CONFIGURATION_PROPERTIES);
 
@@ -162,6 +176,130 @@ public class ConnectorHandlerTest {
         verify(dataManager).delete(Connector.class, "connector-id");
         verify(dataManager).delete(Connection.class, "connection1");
         verify(dataManager).delete(Connection.class, "connection2");
+    }
+
+    @Test
+    public void shouldNotFailToEnrichDynamicPropertiesWithNoResponse() {
+        final ConnectorPropertiesHandler connectorPropertiesHandler = mock(ConnectorPropertiesHandler.class);
+
+        final ConnectorHandler connectorHandler = new ConnectorHandler(dataManager, NO_VERIFIER, NO_CREDENTIALS, NO_INSPECTORS, NO_STATE,
+            NO_ENCRYPTION_COMPONENT, applicationContext, NO_ICON_DAO, NO_EXTENSION_DATA_MANAGER,
+            NO_METADATA_CONFIGURATION_PROPERTIES) {
+            @Override
+            public ConnectorPropertiesHandler properties(@NotNull String connectorId) {
+                return connectorPropertiesHandler;
+            }
+        };
+
+        @SuppressWarnings("resource")
+        final Response metaResponse = responseWithEntity("{}");
+
+        when(connectorPropertiesHandler.enrichWithDynamicProperties("connectorId", null)).thenReturn(metaResponse);
+
+        final Connector connector = new Connector.Builder()
+            .id("connectorId")
+            .build();
+        final Connector withDynamicProperties = connectorHandler.enrichWithDynamicProperties(connector);
+
+        final Connector expected = new Connector.Builder()
+            .id("connectorId")
+            .build();
+
+        assertThat(withDynamicProperties).isEqualTo(expected);
+    }
+
+    @Test
+    public void shouldEnrichDynamicPropertiesWithResponseFromMeta() {
+        final ConnectorPropertiesHandler connectorPropertiesHandler = mock(ConnectorPropertiesHandler.class);
+
+        final ConnectorHandler connectorHandler = new ConnectorHandler(dataManager, NO_VERIFIER, NO_CREDENTIALS, NO_INSPECTORS, NO_STATE,
+            NO_ENCRYPTION_COMPONENT, applicationContext, NO_ICON_DAO, NO_EXTENSION_DATA_MANAGER,
+            NO_METADATA_CONFIGURATION_PROPERTIES) {
+            @Override
+            public ConnectorPropertiesHandler properties(@NotNull String connectorId) {
+                return connectorPropertiesHandler;
+            }
+        };
+
+        @SuppressWarnings("resource")
+        final Response metaResponse = responseWithEntity("{\"properties\":{\"property\":[{\"displayValue\":\"Value 1\",\"value\":\"value1\"},{\"displayValue\":\"Value 2\",\"value\":\"value2\"}]}}");
+
+        when(connectorPropertiesHandler.enrichWithDynamicProperties("connectorId", null)).thenReturn(metaResponse);
+
+        final Connector connector = new Connector.Builder()
+            .id("connectorId")
+            .putProperty("property", new ConfigurationProperty.Builder().build())
+            .build();
+        final Connector withDynamicProperties = connectorHandler.enrichWithDynamicProperties(connector);
+
+        final Connector expected = new Connector.Builder()
+            .id("connectorId")
+            .putProperty("property", new ConfigurationProperty.Builder()
+                .addEnum(PropertyValue.Builder.of("value1", "Value 1"), PropertyValue.Builder.of("value2", "Value 2"))
+                .build())
+            .build();
+
+        assertThat(withDynamicProperties).isEqualTo(expected);
+    }
+
+    private static Response responseWithEntity(final String data) {
+        final ClientConfiguration configuration = new ClientConfiguration(new ResteasyProviderFactory().register(JacksonJsonProvider.class));
+        final Response metaResponse = spy(new ClientResponse(configuration) {
+            @Override
+            protected InputStream getInputStream() {
+                return new ByteArrayInputStream(data.getBytes(StandardCharsets.US_ASCII));
+            }
+
+            @Override
+            protected void setInputStream(InputStream is) {
+                // nop
+            }
+
+            @Override
+            public void releaseConnection() throws IOException {
+                // nop
+            }
+
+            @Override
+            public void releaseConnection(boolean consumeInputStream) throws IOException {
+                // nop
+            }
+
+            @Override
+            public MediaType getMediaType() {
+                return MediaType.APPLICATION_JSON_TYPE;
+            }
+        });
+        return metaResponse;
+    }
+
+    @Test
+    public void shouldNotFailToEnrichDynamicPropertiesWithErrorResponse() {
+        final ConnectorPropertiesHandler connectorPropertiesHandler = mock(ConnectorPropertiesHandler.class);
+
+        final ConnectorHandler connectorHandler = new ConnectorHandler(dataManager, NO_VERIFIER, NO_CREDENTIALS, NO_INSPECTORS, NO_STATE,
+            NO_ENCRYPTION_COMPONENT, applicationContext, NO_ICON_DAO, NO_EXTENSION_DATA_MANAGER,
+            NO_METADATA_CONFIGURATION_PROPERTIES) {
+            @Override
+            public ConnectorPropertiesHandler properties(@NotNull String connectorId) {
+                return connectorPropertiesHandler;
+            }
+        };
+
+        @SuppressWarnings("resource")
+        final Response metaResponse = Response.serverError().build();
+        when(connectorPropertiesHandler.enrichWithDynamicProperties("connectorId", null)).thenReturn(metaResponse);
+
+        final Connector connector = new Connector.Builder()
+            .id("connectorId")
+            .build();
+        final Connector withDynamicProperties = connectorHandler.enrichWithDynamicProperties(connector);
+
+        final Connector expected = new Connector.Builder()
+            .id("connectorId")
+            .build();
+
+        assertThat(withDynamicProperties).isEqualTo(expected);
     }
 
     private static ConnectorAction newActionBy(final Connector connector) {

--- a/app/server/endpoint/src/test/java/io/syndesis/server/endpoint/v1/handler/connection/ConnectorPropertiesHandlerTest.java
+++ b/app/server/endpoint/src/test/java/io/syndesis/server/endpoint/v1/handler/connection/ConnectorPropertiesHandlerTest.java
@@ -1,0 +1,70 @@
+/*
+ * Copyright (C) 2016 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.syndesis.server.endpoint.v1.handler.connection;
+
+import java.util.Collections;
+import java.util.Map;
+
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.Invocation;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import io.syndesis.server.verifier.MetadataConfigurationProperties;
+
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class ConnectorPropertiesHandlerTest {
+
+    @Test
+    public void shouldSendRequestsToMeta() {
+        final MetadataConfigurationProperties config = new MetadataConfigurationProperties();
+        config.setService("syndesis-meta");
+
+        final Client client = mock(Client.class);
+
+        final ConnectorPropertiesHandler handler = new ConnectorPropertiesHandler(config) {
+            @Override
+            Client createClient() {
+                return client;
+            }
+        };
+
+        final ArgumentCaptor<String> url = ArgumentCaptor.forClass(String.class);
+
+        final WebTarget target = mock(WebTarget.class);
+        when(client.target(url.capture())).thenReturn(target);
+        final Invocation.Builder builder = mock(Invocation.Builder.class);
+        when(target.request()).thenReturn(builder);
+        final Map<String, Object> properties = Collections.emptyMap();
+        @SuppressWarnings("resource")
+        final Response response = mock(Response.class);
+        when(builder.post(Entity.entity(properties, MediaType.APPLICATION_JSON_TYPE))).thenReturn(response);
+
+        @SuppressWarnings("resource")
+        final Response received = handler.enrichWithDynamicProperties("connectorId", properties);
+
+        assertThat(received).isSameAs(response);
+        assertThat(url.getValue()).isEqualTo("http://syndesis-meta/api/v1/connectors/connectorId/properties/meta");
+    }
+}

